File: Data Structures & Algorithms/design-hashmap/submission-3.java
class ListNode{
    int key;
    int value;
    ListNode next;
    
    public ListNode(int key, int value){
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

class MyHashMap {
    private ListNode[] map;

    public MyHashMap() {
        map = new ListNode[10000];

        for(int i = 0; i < 10000; i++){
            map[i] = new ListNode(0,0);
        }
    }
    
    public void put(int key, int value) {
        int idx = key % map.length;
        ListNode cur = map[idx];
        while (cur.next != null){
            if(cur.next.key == key){
                cur.next.value = value;
                return;
            }
            cur = cur.next;
        }
        cur.next = new ListNode(key,value);
    }
    
    public int get(int key) {
        int idx = key % map.length;
        ListNode cur = map[idx];
        while(cur.next != null){
            if(cur.next.key == key){
                return cur.next.value;
            }
            cur = cur.next;
        }
        return -1;
    }
    
    public void remove(int key) {
        int idx = key % map.length;
        ListNode cur = map[idx];
        while(cur.next != null){
            if(cur.next.key == key){
                cur.next= cur.next.next;
                return;
            }
            cur = cur.next;
        }
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

File: Data Structures & Algorithms/range-sum-query-2d-immutable/submission-3.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        row, col= len(matrix), len(matrix[0])
        self.prefixsum= [[0]*col for i in range(row)]
        
        for i in range(row):
            for j in range(col):
                top= self.prefixsum[i-1][j] if i>0 else 0
                left= self.prefixsum[i][j-1] if j>0 else 0
                topleft= self.prefixsum[i-1][j-1] if i>0 and j>0 else 0
                self.prefixsum[i][j]= matrix[i][j]+top+left-topleft        

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        
        total = self.prefixsum[row2][col2]
        left = self.prefixsum[row2][col1-1] if col1>0 else 0
        top = self.prefixsum[row1-1][col2] if row1>0 else 0
        topleft = self.prefixsum[row1-1][col1-1] if row1>0 and col1 > 0 else 0 
      
        return total-left-top+topleft
        

# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)